### ISSUE TITLE
Version 0.4.0

### ISSUE THREAD

**Issue by Gcarmnonapy7** [MorphosML/MorphosML#16]
### Issue Body:

```markdown
# [RFC / Roadmap] MorphosML v0.4.0: Hardware-Accelerated Neural Engine, Deep Autodiff & latexify_py Integration

## 1. Executive Summary & Context

With the release of **MorphosML v0.3.1**, we eliminated our fundamental memory and ingestion bottlenecks:
- **Contiguous 1D flat buffers & zero-copy `TensorView`**: Delivered a **216.6x latency speedup** and **0.0 MB** heap allocation overhead over v0.2.0.
- **Out-of-Core Data Ingestion**: POSIX `mmap` streaming and asynchronous double-buffered `DataLoader` reached **15.87 Million samples/sec**.
- **Calculus Core & Verification**: $\mathcal{O}(h^4)$ numerical differentiation, quadrature integration, and 71/71 tests passing green.

The objective of **v0.4.0** is to transition MorphosML into a unified **"Compute + Differentiate + Typeset"** framework:
1. **Hardware Acceleration**: Exploit CPU silicon to the limit with AVX2/AVX-512 SIMD vectorization and OpenMP multi-threaded parallel GEMM.
2. **Deep Autodiff & Neural Network Core (`morphosml.nn`)**: Implement a tape-based reverse-mode automatic differentiation graph, neural layers (`Linear`, `ReLU`, `GELU`, `Softmax`), and first-class optimizers (`Adam`, `SGD`).
3. **`google/latexify_py` Integration**: Seamlessly compile mathematical functions and neural network architectures into publication-ready LaTeX equations via AST parsing.

---

## 2. System Architecture

```
MorphosML v0.4.0 Topology
├── Hardware Acceleration Layer (C++17 + SIMD + OpenMP)
│   ├── SIMD Kernels           (AVX2 / AVX-512 for x86_64, ARM NEON for aarch64)
│   │   ├── simd_dot_product   (4x–8x vector dot product throughput)
│   │   ├── simd_vector_ops    (Vectorized FMA element-wise addition/multiplication)
│   │   └── simd_gemm          (Cache-blocked SIMD matrix multiplication)
│   └── Multi-Threaded Engine  (OpenMP row-chunk parallelization across all CPU cores)
│
├── Autodiff & Neural Network Core (`morphosml.nn` & `morphosml.autodiff`)
│   ├── Computation Tape       (Dynamic DAG recording operations for backpropagation)
│   ├── Variable & Gradients   (Automatic reverse-mode backpropagation via .backward())
│   ├── Layers (`nn.Module`)   (Linear, ReLU, GELU, Sigmoid, Softmax, Sequential)
│   ├── Loss Functions         (MSELoss, CrossEntropyLoss, BCELoss)
│   └── Optimizers (`optim`)   (Adam with momentum & weight decay, SGD)
│
└── LaTeX Mathematical Compiler (`morphosml.latex` + `google/latexify_py`)
    ├── AST Function Parser    (Direct interop with google/latexify_py for Python formulas)
    ├── NN Architecture LaTeX  (Decompile nn.Sequential into composition LaTeX equations)
    └── Live Notebook Proofs   (KaTeX rendering of computational graphs & loss curves)
```

---

## 3. Scope & Technical Specifications

### Pillar 1: Hardware Acceleration (SIMD & OpenMP Parallelism)
- [ ] **AVX2 / AVX-512 Vectorization**: Implement dedicated vector kernels in `cpp/include/morphosml/simd/` for:
  - Fused Multiply-Add (FMA) dot products (`_mm256_fmadd_pd`).
  - Vector Euclidean norm and normalization.
  - Matrix-vector multiplications ($y = A x$).
- [ ] **ARM NEON Support**: Fallback intrinsics for Apple Silicon and ARM64 servers (`vaddq_f64`, `vmulq_f64`).
- [ ] **Runtime CPU Feature Detection**: Gracefully select optimal instruction sets (AVX-512 $\to$ AVX2 $\to$ scalar fallback) without crashing older hardware.
- [ ] **OpenMP Parallel GEMM**: Cache-blocked $i$-$k$-$j$ matrix multiplication parallelized over CPU threads:
  ```cpp
  #pragma omp parallel for collapse(2) schedule(static)
  ```

### Pillar 2: Reverse-Mode Automatic Differentiation
- [ ] **Computation Tape / DAG**:
  - `morphosml::autodiff::Variable`: Wraps contiguous data buffers with gradient accumulators.
  - Reverse-mode dynamic tape recording operations (`Add`, `Sub`, `Mul`, `MatMul`, `GELU`, `Sigmoid`).
- [ ] **Exact Backpropagation**:
  - `.backward()` computes exact analytical gradients in a single reverse sweep without finite-difference truncation error.

### Pillar 3: Neural Network Engine (`morphosml.nn` & `morphosml.optim`)
- [ ] **Base Class `nn.Module`**:
  - Parameter registration (`.parameters()`), zero-grad (`.zero_grad()`), train/eval mode flags.
- [ ] **Core Layers**:
  - `nn.Linear(in_features, out_features, bias=True)`
  - `nn.ReLU()`, `nn.GELU()`, `nn.Sigmoid()`, `nn.Softmax(dim=-1)`
  - `nn.Sequential(*layers)`
- [ ] **Loss Functions**:
  - `nn.MSELoss()`, `nn.CrossEntropyLoss()`, `nn.BCELoss()`
- [ ] **Optimizers**:
  - `optim.SGD(params, lr=0.01, momentum=0.9)`
  - `optim.Adam(params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8, weight_decay=0.0)`

### Pillar 4: Deep Integration with `google/latexify_py`
- [ ] **AST Decompilation for Mathematical Functions**:
  - Integrate with `latexify-py` so any `@latexify.function` can be passed directly into MorphosML calculus routines.
  - Dual decorators: `@latexify.function` + `@differentiable` yielding both exact symbolic LaTeX and C++ hardware-accelerated derivatives.
- [ ] **Neural Network Architecture to LaTeX**:
  - Add `ml.latex.latexify(model)` support for `nn.Sequential`, decompiling complex multi-layer perceptrons into mathematical equations:
    $$\hat{\mathbf{y}} = \mathrm{softmax}\left(\mathbf{W}_2 \cdot \mathrm{GELU}(\mathbf{W}_1 \mathbf{x} + \mathbf{b}_1) + \mathbf{b}_2\right)$$
- [ ] **Jupyter Notebook Integration**:
  - Rich `_repr_latex_()` display hooks rendering trained models and loss functions directly as interactive KaTeX formulas.

---

## 4. Target API Demonstration

```python
import latexify
import morphosml as ml
import morphosml.nn as nn
import morphosml.optim as optim

# ==========================================================
# 1. Hardware-Accelerated Multi-Layer Perceptron
# ==========================================================
model = nn.Sequential(
    nn.Linear(in_features=4, out_features=16),
    nn.GELU(),
    nn.Linear(in_features=16, out_features=3),
    nn.Softmax()
)

# Render architecture into publication-ready LaTeX math!
print(ml.latex.latexify(model))
# Output: \hat{\mathbf{y}} = \mathrm{softmax}\left(\mathbf{W}_2 \cdot \mathrm{GELU}(\mathbf{W}_1 \mathbf{x} + \mathbf{b}_1) + \mathbf{b}_2\right)

# ==========================================================
# 2. End-to-End Training Loop with C++ SIMD & Adam
# ==========================================================
optimizer = optim.Adam(model.parameters(), lr=1e-3)
criterion = nn.CrossEntropyLoss()

for batch_x, batch_y in dataloader:
    optimizer.zero_grad()
    predictions = model(batch_x)          # AVX2/AVX-512 accelerated forward
    loss = criterion(predictions, batch_y)
    loss.backward()                       # Reverse-mode tape autodiff
    optimizer.step()                      # SIMD vectorized parameter updates

# ==========================================================
# 3. Dual-Mode Calculus with Google latexify_py
# ==========================================================
@latexify.function
@ml.differentiable
def loss_landscape(x, y):
    return (1.0 - x)**2 + 100.0 * (y - x**2)**2

# Formatted LaTeX math formula via latexify_py:
print(loss_landscape)

# Exact C++ gradient & curvature:
print("Gradient:", loss_landscape.grad([1.0, 1.0]))
print("Hessian: ", loss_landscape.hessian([1.0, 1.0]))
```

---

## 5. Non-Functional Requirements & Acceptance Criteria

- **Standards**: Modern C++17, Python $\ge$ 3.10.
- **Safety**: 0 memory leaks (validated via AddressSanitizer and Valgrind).
- **Quality**: 100% passing test suite with Black formatting and Ruff PEP 8 compliance.
- **Backwards Compatibility**: Zero breaking changes to existing `Matrix`, `Vector`, `TensorView`, `MMapDataset`, and `DataLoader` APIs.
- **Benchmarking**: Automated benchmark script comparing v0.3.1 scalar baseline with v0.4.0 SIMD/OpenMP throughput.

---

## 6. Phased Implementation Roadmap

- [ ] **Phase 1: SIMD & Multi-Threading** (AVX2/AVX-512 FMA kernels + OpenMP GEMM).
- [ ] **Phase 2: Autodiff Engine** (`Variable`, dynamic computation tape, reverse-mode `.backward()`).
- [ ] **Phase 3: Neural Network Modules & Optimizers** (`nn.Linear`, activations, `Adam`, `SGD`).
- [ ] **Phase 4: `latexify_py` Integration** (AST parsing, `nn.Sequential` LaTeX formatting, Jupyter hooks).
- [ ] **Phase 5: Benchmarking, Documentation & Release v0.4.0**.
```
